middleware/etcd/lookup_test.go

```diff
@@ -0,0 +1,393 @@
+package etcd
+
+// etcd needs to be running on http://127.0.0.1:2379
+// *and* needs connectivity to the internet for remotely resolving
+// names.
+
+import (
+	"encoding/json"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/miekg/coredns/middleware"
+	"github.com/miekg/coredns/middleware/etcd/msg"
+	"github.com/miekg/coredns/middleware/etcd/singleflight"
+	"github.com/miekg/coredns/middleware/proxy"
+	"github.com/miekg/dns"
+
+	etcdc "github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+)
+
+var (
+	etc    Etcd
+	client etcdc.KeysAPI
+	ctx    context.Context
+)
+
+type Section int
+
+const (
+	Answer Section = iota
+	Ns
+	Extra
+)
+
+func init() {
+	ctx = context.TODO()
+
+	etcdCfg := etcdc.Config{
+		Endpoints: []string{"http://localhost:2379"},
+	}
+	cli, _ := etcdc.New(etcdCfg)
+	etc = Etcd{
+		Proxy:      proxy.New([]string{"8.8.8.8:53"}),
+		PathPrefix: "skydns",
+		Ctx:        context.Background(),
+		Inflight:   &singleflight.Group{},
+		Zones:      []string{"skydns.test."},
+		Client:     etcdc.NewKeysAPI(cli),
+	}
+}
+
+func set(t *testing.T, e Etcd, k string, ttl time.Duration, m *msg.Service) {
+	b, err := json.Marshal(m)
+	if err != nil {
+		t.Fatal(err)
+	}
+	path, _ := e.PathWithWildcard(k)
+	e.Client.Set(ctx, path, string(b), &etcdc.SetOptions{TTL: ttl})
```

Comment:
everywhere here you want to use a `context.WithTimeout(X * time.Millisecond, parentContext)`

This will make sure that your code doesn't hang in case etcd is slow